x/oracle/keeper/native_token.go

```diff
@@ -0,0 +1,291 @@
+package keeper
+
+import (
+	"errors"
+	"fmt"
+
+	sdkmath "cosmossdk.io/math"
+	"github.com/ExocoreNetwork/exocore/x/oracle/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// deposit: update staker's totalDeposit
+// withdoraw: update staker's totalDeposit
+// delegate: update operator's price, operator's totalAmount, operator's totalShare, staker's share
+// undelegate: update operator's price, operator's totalAmount, operator's totalShare, staker's share
+// msg(refund or slash on beaconChain): update staker's price, operator's price
+
+const maxEffectiveBalance = 32
+
+var stakerList types.StakerList
+
+// GetStakerInfo returns details about staker for native-restaking under asset of assetID
+func (k Keeper) GetStakerInfo(ctx sdk.Context, assetID, stakerAddr string) types.StakerInfo {
+	store := ctx.KVStore(k.storeKey)
+	stakerInfo := types.StakerInfo{}
+	value := store.Get(types.NativeTokenStakerKey(assetID, stakerAddr))
+	if value == nil {
+		return stakerInfo
+	}
+	k.cdc.MustUnmarshal(value, &stakerInfo)
+	return stakerInfo
+}
+
+func (k Keeper) UpdateNativeTokenValidatorListForStaker(ctx sdk.Context, chainID, stakerAddr, validatorPubkey string, amount sdkmath.Int) error {
+	// TODO: use definition from assets module
+	assetID := "0xe_" + chainID
+	// emit an event to tell that a staker's validator list has changed
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeCreatePrice,
+		sdk.NewAttribute(types.AttributeKeyNativeTokenUpdate, types.AttributeValueNativeTokenUpdate),
+	))
+	store := ctx.KVStore(k.storeKey)
+	key := types.NativeTokenStakerKey(assetID, stakerAddr)
+	stakerInfo := &types.StakerInfo{}
+	if value := store.Get(key); value == nil {
+		// create a new item for this staker
+		stakerInfo = types.NewStakerInfo(stakerAddr, validatorPubkey)
+	} else {
+		k.cdc.MustUnmarshal(value, stakerInfo)
+		if amount.IsPositive() {
+			// deopsit add a new validator into staker's validatorList
+			stakerInfo.ValidatorPubkeyList = append(stakerInfo.ValidatorPubkeyList, validatorPubkey)
+		}
+	}
+
+	newBalance := types.BalanceInfo{}
+
+	if latestIndex := len(stakerInfo.BalanceList) - 1; latestIndex >= 0 {
+		newBalance = *(stakerInfo.BalanceList[latestIndex])
+		newBalance.Index++
+	}
+	//	newBalance := *(stakerInfo.BalanceList[latestIndex])
+	//	newBalance.Index++
+	newBalance.Block = uint64(ctx.BlockHeight())
+	if amount.IsPositive() {
```

Comment:
Is there a validity check for the **amount** value range?

Reply:
this is called through from assets module, and assets module is responsible for the validity check